// @edgio/core/runtime/edge-functions/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SEMAPHORE_VALUE_SHOULD_WAKE = exports.SEMAPHORE_VALUE_SHOULD_WAIT = exports.SEMAPHORE_INDEX_WORKER_WAIT = exports.eventToString = exports.EVENT_IO_RESPONSE_CHUNK = exports.EVENT_IO_FETCH = exports.EVENT_ERROR = void 0;
// Constants used in EdgeFunction*.ts and sdk.js
// The event constants are copied from sdk/consts.js
exports.EVENT_ERROR = -1;
exports.EVENT_IO_FETCH = 1;
exports.EVENT_IO_RESPONSE_CHUNK = 2;
function eventToString(event) {
    switch (event) {
        case exports.EVENT_ERROR:
            return 'EVENT_ERROR';
        case exports.EVENT_IO_FETCH:
            return 'EVENT_IO_FETCH';
        case exports.EVENT_IO_RESPONSE_CHUNK:
            return 'EVENT_IO_RESPONSE_CHUNK';
        default:
            return `Unknown event: ${event}`;
    }
}
exports.eventToString = eventToString;
exports.SEMAPHORE_INDEX_WORKER_WAIT = 0;
exports.SEMAPHORE_VALUE_SHOULD_WAIT = 0;
exports.SEMAPHORE_VALUE_SHOULD_WAKE = 1;
