// @edgio/core/runtime/image-optimizer/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAX_IMAGE_SIZE = exports.SUPPORTED_CONTENT_TYPES = exports.SUPPORTED_EXTENSIONS = exports.SUPPORTED_TYPES = exports.FIT_OPTIONS = exports.AUTO_OPTIONS = void 0;
exports.AUTO_OPTIONS = {
    webp: 'webp',
    saveData: 'save-data',
    smallest: 'smallest',
};
exports.FIT_OPTIONS = {
    crop: 'crop',
    cover: 'cover',
    bounds: 'bounds',
    smart: 'smart',
    default: 'default',
};
/**
 * Those content types that are supported by Image Optimizer.
 */
exports.SUPPORTED_TYPES = {
    jpg: 'image/jpeg',
    jpeg: 'image/jpeg',
    pjpg: 'image/pjpeg',
    pjpeg: 'image/pjpeg',
    png: 'image/png',
    ppng: 'image/ppng',
    gif: 'image/gif',
    webp: 'image/webp',
    tiff: 'image/tiff',
    tif: 'image/tif',
};
exports.SUPPORTED_EXTENSIONS = Object.keys(exports.SUPPORTED_TYPES);
exports.SUPPORTED_CONTENT_TYPES = Object.values(exports.SUPPORTED_TYPES);
exports.MAX_IMAGE_SIZE = 10 * 1024 * 1024; // 10 MiB
