// @edgio/core/lambda/global.helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EdgioRuntimeGlobal = void 0;
String.prototype.writeToStream = function (response) {
    response.sendResponse(this.toString());
};
/**
 * This class is used to store global variables that are used in the runtime,
 * but are not passed as parameters to the functions, instead they are stored
 * in the global scope. For instance, as we don't have a way to pass values
 * between prod.js or dev.js script, we do it here, by accesing the global scope.
 */
class EdgioRuntimeGlobal {
    static get runtimeOptions() {
        // @ts-ignore
        return global.EDGIO_RUNTIME_OPTIONS;
    }
    static set runtimeOptions(options) {
        // @ts-ignore
        global.EDGIO_RUNTIME_OPTIONS = options;
    }
    static get isCacheEnabled() {
        var _a;
        // @ts-ignore
        return (_a = global.isCacheEnabled) !== null && _a !== void 0 ? _a : false;
    }
    static set isCacheEnabled(value) {
        // @ts-ignore
        global.isCacheEnabled = value;
    }
    static get devOptions() {
        // @ts-ignore
        return global.devOptions;
    }
    static set devOptions(options) {
        // @ts-ignore
        global.devOptions = options;
    }
}
exports.EdgioRuntimeGlobal = EdgioRuntimeGlobal;
