// @edgio/devtools/addBuiltInRoutes.js
const path = require('path')

const FAR_FUTURE_TTL = 60 * 60 * 24 * 365 * 10

const PUBLIC_CACHE_CONFIG = {
  edge: {
    maxAgeSeconds: FAR_FUTURE_TTL,
  },
  browser: false,
}

const EDGIO_DEVTOOLS_ENABLED_COOKIE = 'edgio_devtools_enabled'
const EDGIO_DEVTOOLS_ENV_ENABLED_COOKIE = 'edgio_devtools_env_enabled'

function isDevtoolEnabledByDefault() {
  if (process.env.EDGIO_DEVTOOLS_ENABLED) {
    return process.env.EDGIO_DEVTOOLS_ENABLED === 'true'
  }
  // EDGIO_DEVTOOLS_ENABLED cannot be set manually on Environment config
  // because EDGIO_* is reserved for internal usage, which we will use when adding
  // that feature explictely to LD interface.
  // Meanwhile, the PREVIEW_EDGIO_DEVTOOLS_ENABLED can be set manually by anyone
  if (process.env.PREVIEW_EDGIO_DEVTOOLS_ENABLED) {
    return process.env.PREVIEW_EDGIO_DEVTOOLS_ENABLED === 'true'
  }

  // If neither of those are defined, we fallback on default behavior:
  // it will be enabled on local and moovweb-edge(-dev).io domains
  return null
}

module.exports = function addBuiltInRoutes({ router, local }) {
  if (isDevtoolEnabledByDefault() !== null) {
    router.match('/(.*)', ({ addResponseCookie }) => {
      addResponseCookie(EDGIO_DEVTOOLS_ENV_ENABLED_COOKIE, `${isDevtoolEnabledByDefault()}; Path=/`)
    })
  }
  if (local) {
    router
      .get('/__edgio__/caching', ({ send }) => {
        send(JSON.stringify({ local, enabled: process.env.EDGIO_CACHE === 'true' }))
      })
      .post('/__edgio__/caching', ({ compute, send }) => {
        compute(req => {
          if (req.query) {
            process.env.EDGIO_CACHE = req.query.enabled
          }
        })
        send('done')
      })
  }
  router
    .match('/__edgio__/devtools/:path*', ({ cache, serveStatic }) => {
      cache(PUBLIC_CACHE_CONFIG)
      serveStatic(path.join('node_modules', '@edgio', 'devtools', 'widget', ':path*'))
    })
    .get('/__edgio__/devtools/enable', ({ redirect, addResponseCookie, cache }) => {
      addResponseCookie(EDGIO_DEVTOOLS_ENABLED_COOKIE, 'true; Path=/')
      cache({
        browser: false,
        edge: false,
      })
      redirect('/')
    })
    .get('/__edgio__/devtools/disable', ({ redirect, addResponseCookie, cache }) => {
      addResponseCookie(EDGIO_DEVTOOLS_ENABLED_COOKIE, 'false; Path=/')
      cache({
        browser: false,
        edge: false,
      })
      redirect('/')
    })
    .match('/:path*', {
      headers: {
        debug_header: true,
        // We can't set the debug headers, because x-ec is reserved
        // set_request_headers: {
        //   'x-ec-debug': 'x-ec-cache,x-ec-check-cacheable,x-ec-cache-key,x-ec-cache-state',
        // },
      },
    })
}
