// ts-fs-utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Name = exports.RelativeFolderName = exports.AbsoluteFolderName = exports.FileExtension = exports.FileName = exports.RelativeFileName = exports.AbsoluteFileName = void 0;
var AbsoluteFileName_1 = require("./AbsoluteFileName");
Object.defineProperty(exports, "AbsoluteFileName", { enumerable: true, get: function () { return AbsoluteFileName_1.AbsoluteFileName; } });
var RelativeFileName_1 = require("./RelativeFileName");
Object.defineProperty(exports, "RelativeFileName", { enumerable: true, get: function () { return RelativeFileName_1.RelativeFileName; } });
var FileName_1 = require("./FileName");
Object.defineProperty(exports, "FileName", { enumerable: true, get: function () { return FileName_1.FileName; } });
var FileExtension_1 = require("./FileExtension");
Object.defineProperty(exports, "FileExtension", { enumerable: true, get: function () { return FileExtension_1.FileExtension; } });
var AbsoluteFolderName_1 = require("./AbsoluteFolderName");
Object.defineProperty(exports, "AbsoluteFolderName", { enumerable: true, get: function () { return AbsoluteFolderName_1.AbsoluteFolderName; } });
var RelativeFolderName_1 = require("./RelativeFolderName");
Object.defineProperty(exports, "RelativeFolderName", { enumerable: true, get: function () { return RelativeFolderName_1.RelativeFolderName; } });
var Name_1 = require("./Name");
Object.defineProperty(exports, "Name", { enumerable: true, get: function () { return Name_1.Name; } });
