// @edgio/core/errors/EdgioHeadersOverflowError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EdgioHeadersOverflowError = void 0;
const EdgioError_1 = require("./EdgioError");
const edgioErrorLinks_1 = require("./edgioErrorLinks");
/*
  This error is triggered when default Node's value (80KB) for the maximum size of the response headers is exceeded.
  See https://github.com/nodejs/http-parser/blob/678a9e21f11e5ea93e5e17a2302390f88a6a8c46/http_parser.h#L64
 */
class EdgioHeadersOverflowError extends EdgioError_1.EdgioError {
    constructor(details = 'HPE_HEADER_OVERFLOW', stack) {
        super({
            statusCode: 542,
            title: 'Edgio Headers Overflow',
            message: `The project's cloud function returned HTTP response headers with an overall size greater than 80KB or single header with value greater than 32KB. If you are the website administrator, please use ${edgioErrorLinks_1.serverLogsLink} to debug and decrease the response headers size. See: ${edgioErrorLinks_1.limitsLink}`,
            level: 1,
            details,
            stack,
        });
    }
}
exports.EdgioHeadersOverflowError = EdgioHeadersOverflowError;
